include/awkward/kernels/kernel.h

```diff
@@ -0,0 +1,21 @@
+#ifndef AWKWARD_KERNEL_H
+#define AWKWARD_KERNEL_H
+
+#include "awkward/cpu-kernels/util.h"
+
+class KernelCore {
+public:
+    KernelCore() {}
+
+    static ERROR new_identities32(uint8_t memory_loc,
+                                  int32_t *toptr,
+                                  int64_t length);
+
+    static ERROR new_identities64(uint8_t memory_loc,
+                                  int64_t *toptr,
+                                  int64_t length);
+
+};
+
+
+#endif //AWKWARD_KERNEL_H
```

Comment:
Since it's all going to be contained in this one file, we don't need a `kernels` directory. As part of the C++ library, the file should be named after the one class it contains. Also because there's only one class, the word "Core" is unnecessary.

So how about files named include/awkward/Kernel.h and src/libawkward/Kernel.cpp defining a class named Kernel?

We should keep in mind that this class is part of the C++ libawkward.so, not the C libawkward-cpu-kernels.so or libawkward-gpu-kernels. One of the features it provides is re-templatizing C functions named

   * `awkward_listarray32_getitem_XYZ`
   * `awkward_listarrayU32_getitem_XYZ`
   * `awkward_listarray64_getitem_XYZ`

into a single static method named

   * `Kernel::listarray_getitem_XYZ<T>`

which is something that all of those free functions in `util::` are currently doing (without removing the `awkward_` prefix, but the `Kernel::` class lets us do that).

Also worth considering: if `Kernel` only contains static methods, then it can be a namespace, `kernel::` (lowerspace because it's not a class, and the file name would follow suit). What do you think of that? Do you see any advantages/disadvantages of it being a class?

Reply:
I think a simple namespace would work here since what we are trying to achieve here(a centralized method for dispatching `awkward_cpu_` and `awkward_gpu_` functions and removing all possibilities of function name duplicates) is basically what a `namespace` does. Introducing an additional class brings in the complexity/overhead of a pointer which as discussed above could be mitigated with the use of a `unique_ptr`, but even then it seems unnecessary.

Also, it does make sense to move all those "free" functions into the kernel namespace, such that these newly re-templatized functions dispatch the appropriate function template based on the `memory_loc` integer.